src/main/java/org/apache/datasketches/kll/BaseKllSketch.java

```diff
@@ -0,0 +1,324 @@
+  static final int K_SHORT                = 4;  // to 5
+  static final int M_BYTE                 = 6;
+  //                                        7 is reserved for future use
+  // SINGLE ITEM ONLY
+  static final int DATA_START_SINGLE_ITEM = 8;
+
+  // MULTI-ITEM
+  static final int N_LONG                 = 8;  // to 15
+  static final int MIN_K_SHORT            = 16; // to 17
+  static final int NUM_LEVELS_BYTE        = 18;
+
+  // FLOAT SKETCH                           19 is reserved for future use in float sketch
+  static final int DATA_START_FLOAT       = 20; // float sketch, not single item
+
+  // DOUBLE SKETCH                          19 to 23 is reserved for future use in double sketch
+  static final int DATA_START_DOUBLE      = 24; // double sketch, not single item
+
+  // Other static values
+  static final byte SERIAL_VERSION            = 1;
+  static final byte SERIAL_VERSION_SINGLE     = 2; // only used to specify the single-item format
+  static final int PREAMBLE_INTS_EMPTY_SINGLE = 2; // for empty and single item
+  static final int PREAMBLE_INTS_FLOAT        = 5; // not empty or single item
+  static final int PREAMBLE_INTS_DOUBLE       = 6; // not empty or single item
+
+  enum Flags { IS_EMPTY, IS_LEVEL_ZERO_SORTED, IS_SINGLE_ITEM }
+
+  /*
+   * Data is stored in items_.
+   * The data for level i lies in positions levels_[i] through levels_[i + 1] - 1 inclusive.
+   * Hence levels_ must contain (numLevels_ + 1) indices.
+   * The valid portion of items_ is completely packed, except for level 0.
+   * Level 0 is filled from the top down.
+   *
+   * Invariants:
+   * 1) After a compaction, or an update, or a merge, all levels are sorted except for level zero.
+   * 2) After a compaction, (sum of capacities) - (sum of items) >= 1,
+   *  so there is room for least 1 more item in level zero.
+   * 3) There are no gaps except at the bottom, so if levels_[0] = 0,
+   *  the sketch is exactly filled to capacity and must be compacted.
+   */
+
+  final int k_; // configured value of K
+  final int m_; // configured minimum buffer "width", Must always be DEFAULT_M for now.
+
+  int minK_;      // for error estimation after merging with different k
+  long n_;        // number of items input into this sketch
+  int numLevels_; // one-based number of current levels,
+  int[] levels_;  // array of index offsets into the items[]. Size = numLevels + 1.
+  boolean isLevelZeroSorted_;
+
+  final boolean compatible; //compatible with quantiles sketch
+  static final Random random = new Random();
+
+  /**
+   * Heap constructor.
+   * @param k configured size of sketch. Range [m, 2^16]
+   * @param m minimum level size. Default is 8.
+   */
+  BaseKllSketch(final int k, final int m, final boolean compatible) {
+    checkK(k);
+    k_ = k;
+    minK_ = k;
+    m_ = m;
+    numLevels_ = 1;
+    levels_ = new int[] {k, k};
+    isLevelZeroSorted_ = false;
+    this.compatible = compatible;
+  }
```

Comment:
MIN_K = DEFAULT_M = 8.  This is an artifact of the design.  In theory, this allows the possibility that in the future we might choose MIN_K > DEFAULT_M.  And the possibility that in the future, we could change the value of DEFAULT_M.  In fact, we could probably substitute DEFAULT_M for _m_ everywhere in the code because we are not allowing it to be anything else.